File: firmware/esp32_sensor_node/StateManager.cpp
#include "StateManager.h"

const char* stateToString(DeviceState s) {
    switch (s) {
        case DeviceState::INIT:           return "INIT";
        case DeviceState::CONNECTING:     return "CONNECTING";
        case DeviceState::READY:          return "READY";
        case DeviceState::MONITORING:     return "MONITORING";
        case DeviceState::ALERT:          return "ALERT";
        case DeviceState::ERROR:          return "ERROR";
        case DeviceState::ENROLLING:      return "ENROLLING";
        case DeviceState::AUTHENTICATING: return "AUTHENTICATING";
        case DeviceState::AUTHENTICATED:  return "AUTHENTICATED";
        case DeviceState::REJECTED:       return "REJECTED";
        default:                          return "UNKNOWN";
    }
}

StateManager::StateManager()
    : _state(DeviceState::INIT),
      _previous(DeviceState::INIT),
      _enteredAt(millis()),
      _transitionCount(0) {}

bool StateManager::transition(DeviceState next) {
    if (!_isLegal(_state, next)) {
        Serial.printf("[FSM] ILLEGAL transition %s → %s\n",
                      stateToString(_state), stateToString(next));
        return false;
    }
    _logTransition(_state, next);
    _previous        = _state;
    _state           = next;
    _enteredAt       = millis();
    _transitionCount++;
    return true;
}

unsigned long StateManager::timeInState() const {
    return millis() - _enteredAt;
}

// ── Legal transition table ────────────────────────────────────────────────────
// Encoding: from any state you can always go to ERROR (fault recovery).
// Otherwise only the listed paths are valid.
bool StateManager::_isLegal(DeviceState from, DeviceState to) const {
    if (to == DeviceState::ERROR) return true;   // fault can happen anywhere

    switch (from) {
        case DeviceState::INIT:
            return to == DeviceState::CONNECTING;

        case DeviceState::CONNECTING:
            return to == DeviceState::READY;

        case DeviceState::READY:
            return to == DeviceState::MONITORING;

        case DeviceState::MONITORING:
            return to == DeviceState::MONITORING    ||  // self (no-op cycle)
                   to == DeviceState::ALERT         ||  // anomaly detected
                   to == DeviceState::CONNECTING    ||  // WiFi lost
                   to == DeviceState::ENROLLING     ||  // enrollment triggered
                   to == DeviceState::AUTHENTICATING;   // auth triggered

        case DeviceState::ENROLLING:
            return to == DeviceState::MONITORING;       // done or timed out

        case DeviceState::AUTHENTICATING:
            return to == DeviceState::AUTHENTICATED ||
                   to == DeviceState::REJECTED;

        case DeviceState::AUTHENTICATED:
            return to == DeviceState::ALERT       ||   // anomaly follow-up
                   to == DeviceState::MONITORING;      // display timeout

        case DeviceState::REJECTED:
            return to == DeviceState::MONITORING;

        case DeviceState::ALERT:
            return to == DeviceState::MONITORING;

        case DeviceState::ERROR:
            return to == DeviceState::INIT;

        default:
            return false;
    }
}

void StateManager::_logTransition(DeviceState from, DeviceState to) const {
    Serial.printf("[FSM] %s → %s  (after %lums, transition #%u)\n",
                  stateToString(from), stateToString(to),
                  timeInState(), _transitionCount + 1);
}

File: firmware/rpi_sensor_node/state_manager.py
from enum import Enum
import time


class DeviceState(Enum):
    INIT           = "INIT"
    CONNECTING     = "CONNECTING"
    READY          = "READY"
    MONITORING     = "MONITORING"
    ALERT          = "ALERT"
    ERROR          = "ERROR"
    ENROLLING      = "ENROLLING"
    AUTHENTICATING = "AUTHENTICATING"
    AUTHENTICATED  = "AUTHENTICATED"
    REJECTED       = "REJECTED"


_LEGAL = {
    DeviceState.INIT:           {DeviceState.CONNECTING, DeviceState.ERROR},
    DeviceState.CONNECTING:     {DeviceState.READY, DeviceState.MONITORING, DeviceState.ERROR},
    DeviceState.READY:          {DeviceState.MONITORING, DeviceState.ERROR},
    DeviceState.MONITORING:     {DeviceState.AUTHENTICATING, DeviceState.ENROLLING,
                                 DeviceState.ALERT, DeviceState.CONNECTING, DeviceState.ERROR},
    DeviceState.AUTHENTICATING: {DeviceState.AUTHENTICATED, DeviceState.REJECTED, DeviceState.ERROR},
    DeviceState.AUTHENTICATED:  {DeviceState.MONITORING, DeviceState.ALERT, DeviceState.ERROR},
    DeviceState.REJECTED:       {DeviceState.MONITORING, DeviceState.ALERT, DeviceState.ERROR},
    DeviceState.ENROLLING:      {DeviceState.MONITORING, DeviceState.ERROR},
    DeviceState.ALERT:          {DeviceState.MONITORING, DeviceState.ERROR},
    DeviceState.ERROR:          set(),
}


class StateManager:
    def __init__(self):
        self._state = DeviceState.INIT
        self._entered_at = time.monotonic()

    def transition(self, new_state: DeviceState) -> bool:
        if new_state not in _LEGAL.get(self._state, set()):
            print(f"[FSM] Illegal transition {self._state.value} → {new_state.value}")
            return False
        elapsed_ms = (time.monotonic() - self._entered_at) * 1000
        print(f"[FSM] {self._state.value} → {new_state.value}  ({elapsed_ms:.0f} ms in prev state)")
        self._state = new_state
        self._entered_at = time.monotonic()
        return True

    def current(self) -> DeviceState:
        return self._state

    def time_in_state_ms(self) -> float:
        return (time.monotonic() - self._entered_at) * 1000
